Failing written-after-open tests

**Describe the bug**
When I run tests from this repo, two that seem to be about written-after-open conditions fail:
```
--- FAIL: Test_fdReaddir_opened_file_written (0.00s)
    require.go:331: unexpected value
        expected:
        	[]byte{0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x5, 0x6b, 0xee, 0x1, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0x0, 0x3, 0x0, 0x0, 0x0, 0x2e, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x2, 0x0, 0x0, 0x0, 0x3, 0x0, 0x0, 0x0, 0x2e, 0x2e, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x6, 0x6b, 0xee, 0x1, 0x0, 0x0, 0x0, 0x0, 0x4, 0x0, 0x0, 0x0, 0x4, 0x0, 0x0, 0x0, 0x66, 0x69, 0x6c, 0x65}
        was:
        	[]byte{0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x5, 0x6b, 0xee, 0x1, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0x0, 0x3, 0x0, 0x0, 0x0, 0x2e, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x2, 0x0, 0x0, 0x0, 0x3, 0x0, 0x0, 0x0, 0x2e, 0x2e}
        
        /root/wazero/imports/wasi_snapshot_preview1/fs_test.go:5141
FAIL
FAIL	github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1	31.825s

--- FAIL: TestDirFS_Readdir (0.00s)
    --- FAIL: TestDirFS_Readdir/Sees_files_written_after_open (0.00s)
        require.go:331: expected 1, but was 0
            /root/wazero/internal/sysfs/dirfs_test.go:698
--- FAIL: TestStat (0.00s)
    --- FAIL: TestStat/empty_dir (0.00s)
        require.go:331: expected 2, but was 1: linux
            /root/wazero/internal/sysfs/stat_test.go:37
    --- FAIL: TestStat/not_empty_dir (0.00s)
        require.go:331: expected 3, but was 1: linux
            /root/wazero/internal/sysfs/stat_test.go:63
FAIL
FAIL	github.com/tetratelabs/wazero/internal/sysfs	0.857s
```

**To Reproduce**
I tried this in a Fedora container:
```
$ podman run --rm -it fedora:rawhide
# dnf update -y
# dnf install -y go git
# git clone https://github.com/tetratelabs/wazero
# cd wazero
# make test
```
and also an Ubuntu container:
```
$ podman run --rm -it ubuntu:24.04
# apt update
# apt upgrade -yy
# apt install -yy golang git make
# git clone https://github.com/tetratelabs/wazero
# cd wazero
# make test
```
Both fail as above.

**Expected behavior**
The tests should all pass, as they (seem to?) do on CI.

**Environment (please complete the relevant information):**
- Go version: 1.23.2
- wazero Version: dc08732e57d575a240774bc64e87cbed59f81c34
- Host architecture: amd64
- Runtime mode: n/a

--- a/internal/sysfs/osfile.go
+++ b/internal/sysfs/osfile.go
@@ -4,7 +4,6 @@ import (
 	"io"
 	"io/fs"
 	"os"
-	"runtime"
 
 	experimentalsys "github.com/tetratelabs/wazero/experimental/sys"
 	"github.com/tetratelabs/wazero/internal/fsapi"
@@ -15,7 +14,15 @@ func newOsFile(path string, flag experimentalsys.Oflag, perm fs.FileMode, f *os.
 	// Windows cannot read files written to a directory after it was opened.
 	// This was noticed in #1087 in zig tests. Use a flag instead of a
 	// different type.
-	reopenDir := runtime.GOOS == "windows"
+	//
+	// As POSIX states, if a file is removed from or added to the directory
+	// after the most recent call to opendir() or rewinddir(), whether a
+	// subsequent call to readdir() returns an entry for that file is unspecified.
+	//
+	// So there is no guarantee that files added after opendir() will be visible
+	// in readdir(). We need to reopendir() to get the new state of the directory
+	// before readdir().
+	reopenDir := true
 	return &osFile{path: path, flag: flag, perm: perm, reopenDir: reopenDir, file: f, fd: f.Fd()}
 }
 

